helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/OptimalAssignment.java

```diff
@@ -0,0 +1,67 @@
+package org.apache.helix.controller.rebalancer.waged.model;
+
+import org.apache.helix.HelixException;
+import org.apache.helix.model.IdealState;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+
+/**
+ * The data model represents the optimal assignment of N replicas assigned to M instances;
+ * It's mostly used as the return parameter of an assignment calculation algorithm; If the algorithm failed to find
+ * optimal assignment given the endeavor, the user could check the failure reasons
+ */
+public class OptimalAssignment {
```

Comment:
I like the idea of this object, but it should be convertible from a cluster model, I think. Since we design that model to keep the pending assignment, converting from a cluster model would be very natural.